app/dossierPrograms/dossierPrograms.controllers.js

```diff
@@ -100,19 +82,129 @@ dossierProgramsModule.controller("dossiersProgramSectionController", [
                     }),
                 },
             ];
-            /**
-        for (i = 0; i < resultStage.programStageSections[0].dataElements.length; ++i) {
-            resultStage.programStageSections[0].dataElements[i] = resultStage.programStageSections[0].dataElements[i].dataElement;
-        }
-        */
+
             addtoTOC($scope.toc, null, toc, "programs");
             return stage;
         }
 
+        /*
+         *  @name createStageWithSections
+         *  @description Display stage sections data elements and add them to the table of contents
+         *  @scope dossiersProgramSectionController
+         */
         function createStageWithSections(stage, toc) {
             addtoTOC($scope.toc, stage.programStageSections, toc, "programs");
             return stage;
         }
+
+        /*
+         *  @name makeSectionsCalcMode
+         *  @description Determine the "Calculation mode" of the stage sections data elements
+         *  @scope dossiersProgramSectionController
+         */
+        function makeSectionsCalcMode(stage, assignedDEArray, hiddenSectionsArray) {
+            stage.programStageSections.forEach(pss => {
+                pss.dataElements.forEach(pssDE => {
+                    assignedDEArray.forEach(adeArray => {
+                        if (adeArray.ids.includes(pssDE.id)) {
+                            pssDE.calcMode = { type: "programRule", name: adeArray.name };
+                        } else if (!pssDE.calcMode) {
+                            pssDE.calcMode = { type: "default" };
+                        }
+                    });
```

Comment:
This kind of stateful imperative logic looks a bit weird, but I see the project is not using lodash at all, so let's keep it like this.

Reply:
lodash is among the dependencies and can be used. For example, `dossierPrograms.controllers.js` uses `_.uniq`.

If it helps, the logic is to check if the _sections_ and the _dataElements_ are affected buy _programRules_ of "_HIDESECTION_" or "_ASSIGN_" type.
* "programRuleActionType" == "ASSIGN" ->  Program Rule
* "programRuleActionType" != "ASSIGN" and "HIDESECTION" == true ->  Other
* "programRuleActionType" != "ASSIGN" and "HIDESECTION" == false -> Direct data entry